utils/appliance/__init__.py

```diff
@@ -2378,20 +2419,65 @@ def pop(self):
 stack = ApplianceStack()
 
 
+def parallel_appliances_loader(appliance_list, global_kwargs):
+    """Instantiate a list of appliances from configuration data.
+
+    Args:
+        appliance_list: List of dictionaries that contain parameters for :py:class:`IPAppliance`
+        global_kwargs: Arguments that will be defined for each appliances. Appliance can override.
+
+    Result:
+        List of :py:class:`IPAppliance`
+    """
+    result = []
+    for appliance_kwargs in appliance_list:
+        kwargs = {}
+        kwargs.update(global_kwargs)
+        kwargs.update(appliance_kwargs)
+        if not kwargs.get('base_url'):
+            raise ValueError('Appliance definition {!r} is missing base_url'.format(kwargs))
+
+        result.append(IPAppliance(**{IPAppliance.CONFIG_MAPPING[k]: v for k, v in kwargs.items()}))
+    return result
+
+
+def load_parallel_appliances_from_config(config):
+    """Backwards-compatible config loader.
+
+    The ``config`` contains some global values and ``appliances`` key which contains a list of dicts
+    that have the same keys as ``IPAppliance.CONFIG_MAPPING``'s keys. If ``appliances`` key is not
+    present, it is assumed it is old-format definition and the whole dict is used as a reference
+    for one single appliance.
+
+    The global values in the root of the dict (in case of ``appliances`` present) have lesser
+    priority than the values in appliance definitions themselves
+
+    Args:
+        config: A dictionary with the configuration
+    """
+    if 'appliances' not in config:
+        # old-style setup
+        warnings.warn(
+            'Your conf.env has old-style base_url', category=DeprecationWarning, stacklevel=2)
+        appliances = [{
+            k: config[k]
+            for k in IPAppliance.CONFIG_MAPPING.keys()
+            if k in config}]
+        global_kwargs = {}
+    else:
+        # new-style setup
+        appliances = config['appliances']
+        global_kwargs = {
+            k: config[k]
+            for k in IPAppliance.CONFIG_MAPPING.keys()
+            if k not in IPAppliance.CONFIG_NONGLOBAL and k in config}
+
+    return parallel_appliances_loader(appliances, global_kwargs)
+
+
 def get_or_create_current_appliance():
     if stack.top is None:
-        base_url = conf.env['base_url']
-        if base_url is None or str(base_url.lower()) == 'none':
-            raise ValueError('No IP address specified! Specified: {}'.format(repr(base_url)))
-        openshift_creds = conf.env.get('openshift', {})
-        db_host = conf.env.get('db_host')
-        if not isinstance(openshift_creds, dict):
-            raise TypeError('The openshift entry in env.yaml must be a dictionary')
-        stack.push(
-            IPAppliance(
-                urlparse(base_url),
-                container=conf.env.get('container'),
-                openshift_creds=openshift_creds, db_host=db_host))
+        stack.push(load_parallel_appliances_from_config(conf.env)[0])
```

Comment:
Only push the first appliance from the config onto the stack if its empty?  We could load all the appliances into the stack here and still return the first?  Thoughts?